src/main/java/com/bookingApp/bookingApp/service/ReservationService.java

```diff
@@ -0,0 +1,181 @@
+        });
+
+        Collections.sort(screeningsDtos, new Comparator<ScreeningsDto>() {
+            @Override
+            public int compare(ScreeningsDto o1, ScreeningsDto o2) {
+                if(o1.getTitle().compareTo(o2.getTitle())==0){
+                    return o1.getDate().compareTo(o2.getDate());
+                }else {
+                    return o1.getTitle().compareTo(o2.getTitle());
+                }
+            }
+        });
+        return screeningsDtos;
+    }
+
+    @Override
+    public ScreeningInfoDto findById(Long id) {
+        ScreeningInfoDto s = new ScreeningInfoDto();
+        Screening screening = screeningRepository.findById(id)
+                .orElseThrow(()->new ScreeningNotFoundException(id));
+        List<SeatDto> seatDtos = new ArrayList<>();
+
+        s.setScreeningId(screening.getId());
+        s.setDate(screening.getScreeningDate());
+        s.setTitle(screening.getMovie().getTitle());
+        s.setRoom(screening.getSeatScreenings().get(0).getSeat().getRoom().getRoomNumber());
+        for (SeatScreening seatScreening: screening.getSeatScreenings()) {
+            SeatDto seatDto = new SeatDto();
+            seatDto.setSeatId(seatScreening.getSeat().getId());
+            seatDto.setSeatNumber(seatScreening.getSeat().getSeatNumber());
+            seatDto.setReserved(seatScreening.isReserved());
+
+            seatDtos.add(seatDto);
+        }
+        s.setSeats(seatDtos);
+
+        return s;
+    }
+
+    @Override
+    @Transactional
+    public ReservationBackDto reserve(ReservationDto reservation, Long screeningId){
+        Screening screening = screeningRepository.findById(screeningId)
+                .orElseThrow(()->new ScreeningNotFoundException(screeningId));
+        double totalCost = 0;
+        LocalDateTime expDate = null;
+
+        Duration duration = Duration.between(LocalDateTime.now(),screening.getScreeningDate());
+        if(duration.isNegative() || duration.getSeconds()*60 < 15){
+            throw new TooLateToReserveException(screeningId);
+        }
+
+        Reservation r = new Reservation();
+        r.setName(reservation.getName());
+        r.setSurname(reservation.getSurname());
+
+        for(ReservationSeatDto seat: reservation.getSeats()){
+            SeatScreening s = screening.getSeatScreenings().stream()
+                    .filter(seatScreening -> seatScreening.getSeat().getSeatNumber() ==seat.getSeatNumber())
+                    .findAny()
+                    .orElseThrow(()-> new SeatNotFoundException(seat.getSeatNumber()));
+            if(s.isReserved()){
+                throw new SeatAlreadyReserved(seat.getSeatNumber());
+            }
+
+            seatScreeningRepository.findById(s.getId())
+                    .map(seatScreening -> {
+                        seatScreening.setReserved(true);
+                        return seatScreeningRepository.save(seatScreening);
+                    });
+
+            switch (seat.getTicketType()){
```

Comment:
co sie stanie jak dodamy nowy typ biletu lub zmienimy ceny?

Reply:
niestety bedzie to trzeba zmienic w kodzie